chore(docs):[#894] add concept for Efficient orchestration of edc negotiations (#913)
eclipse-tractusx/item-relationship-service: docs/concept/#894-Efficient-orchestration-of-edc-negotiations/#894-Efficient-orchestration-of-edc-negotiations.md

> # #894 Efficient orchestration of edc negotiations

Reviewer: In general there is a template: Please use this template in the future to have a common structure for concepts. https://github.com/eclipse-tractusx/item-relationship-service/blob/aeadd38e192525c89d563732e7266d09afc88bd7/docs/concept/TEMPLATE_Concept.md

| Key           | Value                                                                    |
|---------------|--------------------------------------------------------------------------|

> | Creation date | 09.10.2024                                                               |

Reviewer: Please have a look at creation date.

| Ticket Id     | https://github.com/eclipse-tractusx/item-relationship-service/issues/894 |    
| State         | DONE                                                                     | 

## Table of Contents

1. [Decision](#Decision)
2. [Rationale](#Rationale)
3. [Approach](#Approach)
4. [Testing Strategy](#Testing-Strategy)

## Decision

> Implementation of a new EDC Orchestrator within the EDC Client Library for efficient management of negotiations. The

Reviewer: ```suggestion
Implementation of a new EDC Orchestrator within the IRS EDC Client Library for efficient management of negotiations. The
```

> Implementation of a new EDC Orchestrator within the EDC Client Library for efficient management of negotiations. The

Reviewer: IMHO I would always add the prefix irs to the edc library to avoid confusion.

> goal is to prevent duplicate negotiations for assets, thereby ensuring more efficient processing in asset consumption.

Reviewer: The goal is to avoid redundant and unnecessary negotiations, ensuring more efficient asset utilization.
"duplicate" is my opinion  not precise enough

## Rationale

> Each job is executed in isolation. This results in many jobs conducting a multitude of unnecessary contract negotiations

Reviewer: Please add a note which job is ment here :-)

for the same DT Registry Asset in the EDC catalog. In case the same catalog offer is requested in different jobs this
leads to not performant overall executions. Therefore, a concept is required how to handle this in the most efficient
way.

## Approach

A new service “NegotiationOrchestrator” is introduced which is a middle layer between the already existing components
EdcSubmodelClientImpl and ContractNegotiationService, EDCCatalogFacade and EndpointDataReferenceCacheService.

The service handles the following tasks:

> - orchestration of catalog requests by id

Reviewer: ```suggestion
- orchestration of catalog requests by assetId
```

> - orchestration of catalog requests by type

Reviewer: ```suggestion
- orchestration of catalog requests by assetType
```

- orchestration of contract negotiations
- caching of ongoing contract negotiations
- caching of EDR tokens

To prevent too many parallel requests to the EDC, an ExecutorService handles the catalog requests and negotiations with
a defined thread pool.

> To cache the currently ongoing negotiations, a ConcurrentHashMap with EDC AssetIds as key can be used. Due to the

Reviewer: I would recommend describing that in more generic way. I would recommend a caching api with a in the MVP state initial implementation using ConcurrentHashMap and could be later changed to a more robust cache implementation.

asynchronous process within NegotiationOrchestrator, the future EDR can be created and added to the cache as soon as the

> method is called and will be removed only after the negotiation is completed. In addition, a time to live can be added

Reviewer: The requirement for TTL (Time-to-Live) needs to be clearly described as a feature that must be implemented. Currently, it is unclear what exactly needs to be done from a developer’s perspective. Proper cache handling is mandatory, and TTL is an urgent requirement.

To ensure efficient operation, TTL should be configurable to prevent unnecessary application restarts that might otherwise be required to clear the cache. Without this configurability, unexpected situations could arise due to stale or unmanageable cache data. Please refine the implementation to explicitly define how TTL should be handled and ensure it aligns with these requirements.

for cleanup.

> To cache EDR tokens, the EndpointDataReferenceCacheService can be used. The caching service wraps EDR tokens into an

Reviewer: Have you though about compression of tokens?

Reviewer: This is a highly critical path. How can we ensure that this mechanism functions comprehensively and without errors? I see a potential issue where tokens that are no longer valid might still be treated as valid due to failure scenarios within the system. This poses a significant risk to the acceptance of the system, as security mechanisms could be compromised.

Additionally, is it possible to externally manipulate this mechanism to assign tokens a longer validity period than what is actually stored in the EDC? If so, this could introduce severe security vulnerabilities. Please clarify how these risks are mitigated and ensure that the mechanism is robust against such scenarios.

Author: The validation of the Token is still done by the EDC. If IRS could somehow change the validity period, that would be a security issue on EDC side. 
The validity period assigned by IRS is just the tokens own expiry date to simplify token handling during IRS processing. If a Token is wrongly labeled and used for a data exchange, even though it is expired, an authentication error is thrown by the EDC and IRS will retry the entire process, this time with a new token. 
All of this is already implemented as part of EndpointDataReferenceCacheService and not part of this concept.

object with a state indicating if the token is valid (i.e. can be used directly to authenticate the dataplane request)
or expired (i.e. the validityPeriod is expired). After expiration the token can be reused to request a new token. Cache

> eviction is handled independent of token state and will use a configurable TTL to evict cached entries.

Reviewer: Please descibe

Author: What should be described?

![edc-flow-orchestrated.svg](edc-flow-orchestrated.svg)

> See [docs/src/uml-diagrams/runtime-view/edc-flow-orchestrated.puml](https://github.com/eclipse-tractusx/item-relationship-service/blob/main/docs/src/uml-diagrams/runtime-view/edc-flow-orchestrated.puml)

Reviewer: Usage of a table description where you define the remarks according to the numbers in the diagramm

Reviewer: all steps should be described in more detail.

received contract offers and will receive a list of Future EDRs.

The process of NegotiationOrchestrator is left out of the chart, since it is the same as previous.

![edc-flow-orchestrated-ddtr.png](edc-flow-orchestrated-ddtr.svg)

See [docs/src/uml-diagrams/runtime-view/edc-flow-orchestrated-ddtr.puml](https://github.com/eclipse-tractusx/item-relationship-service/blob/main/docs/src/uml-diagrams/runtime-view/edc-flow-orchestrated-ddtr.puml)
for sources.

## Testing Strategy

E2E Tests can be re-used, because the data we receive does not change. The only difference is, that fewer negotiations
are done.

For unit- and integration tests, the Wiremock tests “SubmodelFacadeWiremockTest” and “IrsWireMockIntegrationTest”
provide full control over all participating components and can therefore verify every interaction between them.

These tests allow to verify the that Tokens are reused, negotiations for the same Asset are only conducted once and
simultaneous requests to the DTR are still working.
The communication to EDC does not change so the already existing *Support Classes can be reused.

New scenarios have to be defined which aim to test the parallel execution of negotiations.

> ### Test case: start 2 jobs with the same globalAssetId at the same time

Reviewer: ```suggestion
### Test case: Start two irs jobs with the same globalAssetId at the same time
```

Expected result is that both jobs complete, but for the second job, no new negotiations should be started .

Verify that the EDC endpoints for EDC stubs are only interacted with once by using the verification methods of
WiremockSupport.

Verify that the Cache has been used for the registry and submodel negotiation.

> ### Test case: start 2 jobs with different globalAssetIds at the same time

Reviewer: ```suggestion
### Test case: Start two irs jobs with different globalAssetIds at the same time
```

Expected result is that both jobs complete and both had their full set of negotiations conducted.

Verify that the Cache has been used for the registry negotiation.

> Verify that the submodel negotiations are done for each submodel.****

Reviewer: Please do not use pictures in the future. Use mermaid instead to prevent deviations between diagram and concept.

> Verify that the submodel negotiations are done for each submodel.****

Reviewer: where is the request catalog requests by assetType shown?
